x/payment/keeper/storage_fee_charge.go

```diff
@@ -52,47 +52,100 @@ func (k Keeper) ApplyStreamRecordChanges(ctx sdk.Context, streamRecordChanges []
+		rateChanges = append(rateChanges, *types.NewDefaultStreamRecordChangeWithAddr(sdk.MustAccAddressFromHex(flowChange.ToAddress)).WithRateChange(flowChange.Rate))
+		totalRate = totalRate.Add(flowChange.Rate)
+	}
+	streamRecordChange := types.NewDefaultStreamRecordChangeWithAddr(from).WithRateChange(totalRate.Neg())
+	// storage fee preview
+	if ctx.IsCheckTx() {
+		reserveTime := k.GetParams(ctx).VersionedParams.ReserveTime
+		changeRate := totalRate.Neg()
+		event := &types.EventFeePreview{
+			Account: from.String(),
+			Amount:  changeRate.Mul(sdkmath.NewIntFromUint64(reserveTime)).Abs(),
 		}
-		err = k.UpdateStreamRecord(ctx, streamRecord, streamRecordChange)
-		if err != nil {
-			return fmt.Errorf("apply stream record changes for user failed: %w", err)
+		if changeRate.IsPositive() {
+			event.FeePreviewType = types.FEE_PREVIEW_TYPE_UNLOCKED_FEE
+		} else {
+			event.FeePreviewType = types.FEE_PREVIEW_TYPE_PRELOCKED_FEE
 		}
+		_ = ctx.EventManager().EmitTypedEvents(event)
+	}
+	err := k.UpdateStreamRecord(ctx, streamRecord, streamRecordChange)
+	if err != nil {
+		return fmt.Errorf("apply stream record changes for user failed: %w", err)
+	}
 
-		// update flows
-		deltaFlowCount := k.MergeActiveOutFlows(ctx, from, userFlows.Flows) // deltaFlowCount can be negative
-		streamRecord.OutFlowCount = uint64(int64(streamRecord.OutFlowCount) + int64(deltaFlowCount))
+	// update flows
+	deltaFlowCount := k.MergeActiveOutFlows(ctx, from, userFlows.Flows) // deltaFlowCount can be negative
+	streamRecord.OutFlowCount = uint64(int64(streamRecord.OutFlowCount) + int64(deltaFlowCount))
 
-		k.SetStreamRecord(ctx, streamRecord)
+	k.SetStreamRecord(ctx, streamRecord)
+	err = k.ApplyStreamRecordChanges(ctx, rateChanges)
+	if err != nil {
+		return fmt.Errorf("apply stream record changes failed: %w", err)
 	}
-	err = k.ApplyStreamRecordChanges(ctx, streamRecordChanges)
+	return nil
+}
+
+func (k Keeper) applyFrozenUserFlows(ctx sdk.Context, userFlows types.UserFlows, from sdk.AccAddress, streamRecord *types.StreamRecord) error {
+	forced, _ := ctx.Value(types.ForceUpdateStreamRecordKey).(bool)
+	if !forced {
+		return fmt.Errorf("stream record %s is frozen", streamRecord.Account)
+	}
+
+	// the stream record could be totally frozen, or in the process of resuming
+	var activeOutFlows, frozenOutFlows []types.OutFlow
+	var activeRateChanges []types.StreamRecordChange
+	//var frozenRateChanges []types.StreamRecordChange
+	totalActiveRate, totalFrozenRate := sdk.ZeroInt(), sdk.ZeroInt()
+	for _, flowChange := range userFlows.Flows {
+		outFlow := k.GetOutFlow(ctx, sdk.MustAccAddressFromHex(streamRecord.Account), types.OUT_FLOW_STATUS_FROZEN, sdk.MustAccAddressFromHex(flowChange.ToAddress))
+		if outFlow != nil {
+			frozenOutFlows = append(frozenOutFlows, flowChange)
+			//frozenRateChanges = append(frozenRateChanges, *types.NewDefaultStreamRecordChangeWithAddr(sdk.MustAccAddressFromHex(flowChange.ToAddress)).WithFrozenRateChange(flowChange.Rate))
```

Comment:
Remove useless code.

Reply:
i keep it there for reminder. i will try other way later.